src/osd/osd_types.h

```diff
@@ -4363,6 +4363,13 @@ struct pg_log_t {
     if (rollback_info_trimmed_to > newhead)
       rollback_info_trimmed_to = newhead;
 
+    if(log.empty()) {
+      // hmm, the whole log list is divergent, reset everything
+      head = tail = eversion_t();
+      can_rollback_to = eversion_t();
+      rollback_info_trimmed_to = eversion_t();
```

Comment:
I don't can_rollback_to and rollback_info_trimmed_to here are correct.  These values would imply that the osd can rollback entries to the beginning of time.

Also, I think head is supposed to be last_update in general, so that's probably also wrong?

There are unit tests for this code, can you add one to illustrate the problem?

Reply:
> There are unit tests for this code, can you add one to illustrate the problem?

It's fired by the normal tethulogy run, see the ticket linked: https://tracker.ceph.com/issues/44532

Reply:
> There are unit tests for this code, can you add one to illustrate the problem?

It's fired by the normal teuthology run, see the ticket linked: https://tracker.ceph.com/issues/44532

Reply:
> can_rollback_to and rollback_info_trimmed_to here are correct. These values would imply that the osd can rollback entries to the beginning of time.
> 
> Also, I think head is supposed to be last_update in general, so that's probably also wrong?

@athanatos Can you reference the PGLog::merge_log function for a more detailed picture of the symptom?

From https://github.com/ceph/ceph/blob/master/src/osd/PGLog.cc#L478 and below we'll
reset last_update and crt etc. (by calling skip_can_rollback_to_to_head) properly.

Reply:
Are you certain this case was actually hit in your test run above?

I do see that line in PGLog.  I'm still not following how the assert in that bug could be fixed by this patch.  From what I can see in the current version of rewind_from_head, there's already an assert proving that newhead (and therefore head) must be less than each of the divergent entries after rewind_from_head is complete, so I worry that something else is going on.  Resetting head to eversion_t() would of course solve that specific assert, but it actually does change the meaning of the function significantly, so I'd like a more detailed explanation of what happened and why this is the best fix.

Either way, I'd really like a unit test in TestPGLog.cc if possible.

Reply:
Hrm.  Reading rewind_from_head and merge_log, it looks to me like we succeeded at:

      ceph_assert(e.version > head);

but failed

      ceph_assert(head.version == 0 || e.version.version > head.version);

merge_log selects the version to roll back to in the "extent_on_head?" section by starting at the end of the authoritative log (olog) and stepping backwards until it hits the beginning or finds a log entry iter from such that

      if (from->version <= log.head) {
	lower_bound = std::max(lower_bound, from->version);
	++from;
	break;
      }

Note, version here is the whole eversion_t rather than the version_t version field.

We then pass lower_bound into rewind_from_head and walk the local log in reverse until we find p such that

      if (p->version.version <= newhead.version) {
	/*
	 * look at eversion.version here.  we want to avoid a situation like:
	 *  our log: 100'10 (0'0) m 10000004d3a.00000000/head by client4225.1:18529
	 *  new log: 122'10 (0'0) m 10000004d3a.00000000/head by client4225.1:18529
	 *  lower_bound = 100'9
	 * i.e, same request, different version.  If the eversion.version is > the
	 * lower_bound, we it is divergent.
	 */
	++p;
	divergent.splice(divergent.begin(), log, p, log.end());
	break;
      }

and then set head=newhead (which is lower_bound from above).

After rolling back the divergent entries, we then call add() on the remaining entries from merge_log [from, to).  For one of them, e, e.version > head but e.version.version <= head.version.

I'm not actually sure this is a bug only in the case that the log is empty, and it's not clear to me that rewind_from_head is strictly relevant as it didn't select newhead.  It seems to me that the lower_bound we are selecting is less than from->version, but has a version greater than from->version.version.  It's not clear to me how that could happen unless our local log.tail was a version that later turned out to be divergent, and if so, that would be the bug (can't trim log entries that haven't been fully committed).  Does this match your understanding of the problem?  Do you know how it happened?

Reply:
I'm pulling the logs, I'll try to take a closer look tomorrow.

Reply:
Yeah, the authoritative log in ceph-osd.2.log from the bug has 34'5300, so the log entry which had the version of log.tail=19'5300 must have been forgotten.  That's a big problem -- it means that entry 19'5300 was considered committed but later marked divergent (or never made it to a future osd).

I'll have to look further back tomorrow.

Reply:
> and if so, that would be the bug (can't trim log entries that haven't been fully committed).

By trying to limit pg log length under all circumstances, it is possible for us to trim pg logs past lcod now. So I think that could happen.

See https://github.com/ceph/ceph/pull/23098
and https://github.com/ceph/ceph/pull/23317

Reply:
That's really bad.  19'5300 and 34'5300 might well be different log entries.  With your patch, the pg would presumably have gone active, but the object modified by 19'5300 would have *incorrect contents* after recovery because we wouldn't have added it to the missing set.

AFAICT, the only correct way to ensure a "hard bound" on pg log length would be to throttle incoming requests until the ones at the tail of the log are committed.

Reply:
Fortunately, a quick read of calc_trim_to seems to suggest that we still won't trim past min_last_complete_ondisk (at least in master).

Reply:
I'll ask around about this tomorrow.

Reply:
@athanatos That's the old fashion only meaning to be compatible with luminous.

calc_trim_to_aggressive() is the default version!

Reply:
Oops, I mixed up min_last_complete_ondisk with last_update_on_disk.  The second one is the one that matters here.

Reply:
@xiexingguo Ah.  As sage said above, it would be safe as long as we never trim past last_update_on_disk, but it looks like calc_trim_to_aggressive actually can.  I think that's the bug here.

Reply:
@neha-ojha It looks to me like calc_trim_to_aggressive() can trim past last_update_ondisk.  I think that can generally cause object inconsistencies as log merging won't necessarily be able to find all divergent entries.